File: Julia/vae-full-network.jl
mutable struct VAE
    input_dim::Int
    latent_dim::Int
    hidden_dim::Int
    w_enc
    b_enc
    w_mu
    b_mu
    w_log_var
    b_log_var
    w_dec_h
    b_dec_h
    w_dec_out
    b_dec_out
end

function VAE(input_dim::Int, latent_dim::Int)
    hidden_dim = 256
    w_enc = randn(input_dim, hidden_dim) .* 0.01
    b_enc = zeros(hidden_dim)

    w_mu = randn(hidden_dim, latent_dim) .* 0.01
    b_mu = zeros(latent_dim)
    w_log_var = randn(hidden_dim, latent_dim) .* 0.01
    b_log_var = zeros(latent_dim)

    w_dec_h = randn(latent_dim, hidden_dim) .* 0.01
    b_dec_h = zeros(hidden_dim)
    w_dec_out = randn(hidden_dim, input_dim) .* 0.01
    b_dec_out = zeros(input_dim)

    VAE(input_dim, latent_dim, hidden_dim, w_enc, b_enc, w_mu, b_mu, w_log_var, b_log_var, w_dec_h, b_dec_h, w_dec_out, b_dec_out)
end

function forward(model::VAE, x)
    h_enc = max.(0, x * model.w_enc .+ model.b_enc)
    mu = h_enc * model.w_mu .+ model.b_mu
    log_var = h_enc * model.w_log_var .+ model.b_log_var

    std = exp.(0.5 .* log_var)
    eps = randn(size(mu))
    z = mu .+ std .* eps

    h_dec = max.(0, z * model.w_dec_h .+ model.b_dec_h)
    logits = h_dec * model.w_dec_out .+ model.b_dec_out
    x_recon = 1.0 ./ (1.0 .+ exp.(-logits))

    return (x_recon = x_recon, mu = mu, log_var = log_var)
end

function generate(model::VAE, n_samples::Int)
    z = randn(n_samples, model.latent_dim)
    h_dec = max.(0, z * model.w_dec_h .+ model.b_dec_h)
    logits = h_dec * model.w_dec_out .+ model.b_dec_out
    1.0 ./ (1.0 .+ exp.(-logits))
end

File: R/vae-full-network.R
VAE <- setRefClass(
  "VAE",
  fields = list(
    input_dim = "numeric",
    latent_dim = "numeric",
    hidden_dim = "numeric",
    w_enc = "matrix",
    b_enc = "numeric",
    w_mu = "matrix",
    b_mu = "numeric",
    w_log_var = "matrix",
    b_log_var = "numeric",
    w_dec_h = "matrix",
    b_dec_h = "numeric",
    w_dec_out = "matrix",
    b_dec_out = "numeric"
  ),
  methods = list(
    initialize = function(input_dim, latent_dim) {
      input_dim <<- input_dim
      latent_dim <<- latent_dim
      hidden_dim <<- 256

      w_enc <<- matrix(rnorm(input_dim * hidden_dim, sd = 0.01), nrow = input_dim, ncol = hidden_dim)
      b_enc <<- numeric(hidden_dim)

      w_mu <<- matrix(rnorm(hidden_dim * latent_dim, sd = 0.01), nrow = hidden_dim, ncol = latent_dim)
      b_mu <<- numeric(latent_dim)
      w_log_var <<- matrix(rnorm(hidden_dim * latent_dim, sd = 0.01), nrow = hidden_dim, ncol = latent_dim)
      b_log_var <<- numeric(latent_dim)

      w_dec_h <<- matrix(rnorm(latent_dim * hidden_dim, sd = 0.01), nrow = latent_dim, ncol = hidden_dim)
      b_dec_h <<- numeric(hidden_dim)
      w_dec_out <<- matrix(rnorm(hidden_dim * input_dim, sd = 0.01), nrow = hidden_dim, ncol = input_dim)
      b_dec_out <<- numeric(input_dim)
    },
    forward = function(x) {
      h_enc <- pmax(0, x %*% w_enc + b_enc)
      mu <- h_enc %*% w_mu + b_mu
      log_var <- h_enc %*% w_log_var + b_log_var

      std <- exp(0.5 * log_var)
      eps <- matrix(rnorm(length(mu)), nrow = nrow(mu), ncol = ncol(mu))
      z <- mu + std * eps

      h_dec <- pmax(0, z %*% w_dec_h + b_dec_h)
      logits <- h_dec %*% w_dec_out + b_dec_out
      x_recon <- 1 / (1 + exp(-logits))

      list(x_recon = x_recon, mu = mu, log_var = log_var)
    },
    generate = function(n_samples) {
      z <- matrix(rnorm(n_samples * latent_dim), nrow = n_samples, ncol = latent_dim)
      h_dec <- pmax(0, z %*% w_dec_h + b_dec_h)
      logits <- h_dec %*% w_dec_out + b_dec_out
      1 / (1 + exp(-logits))
    }
  )
)
